File: SensorScope/app/src/main/java/com/example/sensorscope/MainActivity.java
package com.example.sensorscope;

import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity implements SensorEventListener {

    private SensorManager sensorManager;
    private Sensor accelSensor, lightSensor, proximitySensor;

    private TextView tvAccelX, tvAccelY, tvAccelZ;
    private TextView tvLight, tvLightLabel;
    private TextView tvProximity, tvProximityLabel;
    private ProgressBar pbAccel, pbLight;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Initialize views
        tvAccelX = findViewById(R.id.tvAccelX);
        tvAccelY = findViewById(R.id.tvAccelY);
        tvAccelZ = findViewById(R.id.tvAccelZ);
        tvLight = findViewById(R.id.tvLight);
        tvLightLabel = findViewById(R.id.tvLightLabel);
        tvProximity = findViewById(R.id.tvProximity);
        tvProximityLabel = findViewById(R.id.tvProximityLabel);
        pbAccel = findViewById(R.id.pbAccel);
        pbLight = findViewById(R.id.pbLight);

        // Prevent crash if XML missing something
        if (tvAccelX == null || pbAccel == null) {
            return; // avoid crash
        }

        // Sensor Manager
        sensorManager = (SensorManager) getSystemService(SENSOR_SERVICE);

        if (sensorManager != null) {
            accelSensor = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
            lightSensor = sensorManager.getDefaultSensor(Sensor.TYPE_LIGHT);
            proximitySensor = sensorManager.getDefaultSensor(Sensor.TYPE_PROXIMITY);
        }

        // Sensor availability check
        if (accelSensor == null) tvAccelX.setText("Accelerometer not available");
        if (lightSensor == null) tvLight.setText("Light sensor not available");
        if (proximitySensor == null) tvProximity.setText("Proximity sensor not available");

        // Set ProgressBar max (IMPORTANT)
        if (pbAccel != null) pbAccel.setMax(300);
        if (pbLight != null) pbLight.setMax(10000);
    }

    @Override
    protected void onResume() {
        super.onResume();

        if (sensorManager != null) {
            if (accelSensor != null)
                sensorManager.registerListener(this, accelSensor, SensorManager.SENSOR_DELAY_UI);

            if (lightSensor != null)
                sensorManager.registerListener(this, lightSensor, SensorManager.SENSOR_DELAY_UI);

            if (proximitySensor != null)
                sensorManager.registerListener(this, proximitySensor, SensorManager.SENSOR_DELAY_UI);
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (sensorManager != null) {
            sensorManager.unregisterListener(this);
        }
    }

    @Override
    public void onSensorChanged(SensorEvent event) {

        if (event == null || event.sensor == null) return;

        switch (event.sensor.getType()) {

            case Sensor.TYPE_ACCELEROMETER:
                float x = event.values[0];
                float y = event.values[1];
                float z = event.values[2];

                tvAccelX.setText("X: " + String.format("%.3f", x) + " m/s²");
                tvAccelY.setText("Y: " + String.format("%.3f", y) + " m/s²");
                tvAccelZ.setText("Z: " + String.format("%.3f", z) + " m/s²");

                double magnitude = Math.sqrt(x * x + y * y + z * z);
                if (pbAccel != null)
                    pbAccel.setProgress((int) Math.min(Math.max(magnitude * 10, 0), 300));
                break;

            case Sensor.TYPE_LIGHT:
                float lux = event.values[0];

                tvLight.setText("Lux: " + String.format("%.1f", lux));

                if (pbLight != null)
                    pbLight.setProgress((int) Math.min(Math.max(lux, 0), 10000));

                String condition;
                if (lux < 10) condition = "Dark 🌑";
                else if (lux < 100) condition = "Dim 🌘";
                else if (lux < 1000) condition = "Indoor 💡";
                else if (lux < 5000) condition = "Bright ☀️";
                else condition = "Direct Sunlight 🌞";

                tvLightLabel.setText("Condition: " + condition);
                break;

            case Sensor.TYPE_PROXIMITY:
                float dist = event.values[0];

                tvProximity.setText("Distance: " + String.format("%.1f", dist) + " cm");

                String status = (dist < 5f) ? "Object nearby 🟡" : "Clear ✅";
                tvProximityLabel.setText("Status: " + status);
                break;
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
        // Not used
    }
}

File: SensorScope/app/src/main/java/com/example/sensorscope/MainActivity.kt
package com.example.sensorscope

import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Bundle
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import kotlin.math.abs
import kotlin.math.sqrt

class MainActivity : AppCompatActivity(), SensorEventListener {

    private lateinit var sensorManager: SensorManager
    private var accelSensor: Sensor? = null
    private var lightSensor: Sensor? = null
    private var proximitySensor: Sensor? = null

    private lateinit var tvAccelX: TextView
    private lateinit var tvAccelY: TextView
    private lateinit var tvAccelZ: TextView
    private lateinit var tvLight: TextView
    private lateinit var tvLightLabel: TextView
    private lateinit var tvProximity: TextView
    private lateinit var tvProximityLabel: TextView
    private lateinit var pbAccel: ProgressBar
    private lateinit var pbLight: ProgressBar

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        tvAccelX = findViewById(R.id.tvAccelX)
        tvAccelY = findViewById(R.id.tvAccelY)
        tvAccelZ = findViewById(R.id.tvAccelZ)
        tvLight = findViewById(R.id.tvLight)
        tvLightLabel = findViewById(R.id.tvLightLabel)
        tvProximity = findViewById(R.id.tvProximity)
        tvProximityLabel = findViewById(R.id.tvProximityLabel)
        pbAccel = findViewById(R.id.pbAccel)
        pbLight = findViewById(R.id.pbLight)

        sensorManager = getSystemService(SENSOR_SERVICE) as SensorManager
        accelSensor = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
        lightSensor = sensorManager.getDefaultSensor(Sensor.TYPE_LIGHT)
        proximitySensor = sensorManager.getDefaultSensor(Sensor.TYPE_PROXIMITY)

        if (accelSensor == null) tvAccelX.text = "Accelerometer not available"
        if (lightSensor == null) tvLight.text = "Light sensor not available"
        if (proximitySensor == null) tvProximity.text = "Proximity sensor not available"
    }

    override fun onResume() {
        super.onResume()
        accelSensor?.let { sensorManager.registerListener(this, it, SensorManager.SENSOR_DELAY_UI) }
        lightSensor?.let { sensorManager.registerListener(this, it, SensorManager.SENSOR_DELAY_UI) }
        proximitySensor?.let { sensorManager.registerListener(this, it, SensorManager.SENSOR_DELAY_UI) }
    }

    override fun onPause() {
        super.onPause()
        sensorManager.unregisterListener(this)
    }

    override fun onSensorChanged(event: SensorEvent?) {
        event ?: return
        when (event.sensor.type) {
            Sensor.TYPE_ACCELEROMETER -> {
                val x = event.values[0]
                val y = event.values[1]
                val z = event.values[2]
                tvAccelX.text = "X: ${"%.3f".format(x)} m/s²"
                tvAccelY.text = "Y: ${"%.3f".format(y)} m/s²"
                tvAccelZ.text = "Z: ${"%.3f".format(z)} m/s²"
                val magnitude = sqrt(x * x + y * y + z * z)
                pbAccel.progress = (magnitude * 10).toInt().coerceIn(0, 300)
            }
            Sensor.TYPE_LIGHT -> {
                val lux = event.values[0]
                tvLight.text = "Lux: ${"%.1f".format(lux)}"
                pbLight.progress = lux.toInt().coerceIn(0, 10000)
                tvLightLabel.text = "Condition: " + when {
                    lux < 10 -> "Dark 🌑"
                    lux < 100 -> "Dim 🌘"
                    lux < 1000 -> "Indoor 💡"
                    lux < 5000 -> "Bright ☀️"
                    else -> "Direct Sunlight 🌞"
                }
            }
            Sensor.TYPE_PROXIMITY -> {
                val dist = event.values[0]
                tvProximity.text = "Distance: ${"%.1f".format(dist)} cm"
                tvProximityLabel.text = "Status: " + if (dist < 5f) "Object nearby 🟡" else "Clear ✅"
            }
        }
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
}